DataStructures/BinaryHeap.cs

```diff
@@ -0,0 +1,264 @@
+        /// the replacement value depending on how the removed value compares to its
+        /// replacment value.
+        /// </remarks>
+        /// <param name="element">The element to remove from the heap.</param>
+        /// <exception cref="ArgumentException">Thrown if element is not in heap.</exception>
+        public void Remove(T element)
+        {
+            var idx = data.IndexOf(element);
+
+            if (idx == -1)
+            {
+                throw new ArgumentException($"{element} not in heap!");
+            }
+            else
+            {
+                Swap(idx, data.Count - 1);
+                T tmp = data[data.Count - 1];
+                data.RemoveAt(data.Count - 1);
+
+                if (idx < data.Count)
+                {
+                    if (comparer.Compare(tmp, data[idx]) > 0)
+                    {
+                        HeapifyDown(idx);
+                    }
+                    else
+                    {
+                        HeapifyUp(idx);
+                    }
+                }
+            }
+        }
+
+        /// <summary>
+        /// Swap the elements in the heap array at the given indices.
+        /// </summary>
+        /// <param name="idx1">First index.</param>
+        /// <param name="idx2">Second index.</param>
+        private void Swap(int idx1, int idx2)
+        {
+            T tmp = data[idx1];
+            data[idx1] = data[idx2];
+            data[idx2] = tmp;
+        }
+
+        /// <summary>
+        /// Recursive function to restore heap properties.
+        /// </summary>
+        /// <remarks>
+        /// Restores heap property by swapping the element at <paramref name="elemIdx"/>
+        /// with its parent if the element compares greater than its parent.
+        /// </remarks>
+        /// <param name="elemIdx">The element to check with its parent.</param>
+        private void HeapifyUp(int elemIdx)
+        {
+            var parent = (elemIdx - 1) / 2;
+
+            if (parent >= 0 && comparer.Compare(data[elemIdx], data[parent]) > 0)
+            {
+                Swap(elemIdx, parent);
+                HeapifyUp(parent);
+            }
+        }
+
+        /// <summary>
+        /// Recursive function to restore heap properties.
+        /// </summary>
+        /// <remarks>
+        /// Restores heap property by swapping the element at <paramref name="elemIdx"/>
+        /// with the larger of its children.
+        /// </remarks>
+        /// <param name="elemIdx">The element to check with its children.</param>
+        private void HeapifyDown(int elemIdx)
+        {
+            var left = (2 * elemIdx) + 1;
+            var right = (2 * elemIdx) + 2;
+
+            bool leftLargerThanElem = left < Count ? comparer.Compare(data[elemIdx], data[left]) < 0 : false;
+            bool rightLargerThanElem = right < Count ? comparer.Compare(data[elemIdx], data[left]) < 0 : false;
```

Comment:
Codacy found an issue: [Remove the unnecessary Boolean literal(s).](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=6375083)